gsrs-module-substances-core/src/main/java/ix/ginas/utils/validation/validators/TagsValidator.java

```diff
@@ -0,0 +1,159 @@
+    private Boolean removeTagWhenInTagsMissingFromNames;
+
+    @Autowired
+    private SubstanceRepository substanceRepository;
+
+    // Can't figure out how to set property in test; so doing this.
+    public void setAddTagWhenInTagsPresentInNames(Boolean b) {
+        this.addTagWhenInTagsPresentInNames = b;
+    }
+
+    // Can't figure out how to set property in test; so doing this.
+    public void setRemoveTagWhenInTagsMissingFromNames(Boolean b) {
+        this.removeTagWhenInTagsMissingFromNames = b;
+    }
+
+    @Override
+    public void validate(Substance s, Substance objold, ValidatorCallback callback) {
+        log.info("Inside Tags Validator.validate");
+
+        boolean skipRest = false;
+        if (removeTagWhenInTagsMissingFromNames == null) { removeTagWhenInTagsMissingFromNames = false; }
+
+        List<String> tagTermsFromNames = s.extractDistinctTagTermsFromNames();
+        List<String> tagTerms = s.grabTagTerms();
+
+        if (tagTermsFromNames == null) {
+            log.info("tagTermsFromNames is null.");
+            GinasProcessingMessage mes = GinasProcessingMessage
+                .ERROR_MESSAGE("Null value encountered when extracting tag terms from substance names.");
+            callback.addMessage(mes);
+            skipRest = true;
+        }
+        if (tagTerms == null) {
+            log.info("tagTerms is null.");
+            GinasProcessingMessage mes = GinasProcessingMessage
+                    .ERROR_MESSAGE("Null value encountered when getting substance tag terms.");
+            callback.addMessage(mes);
+            skipRest = true;
+        }
+
+        if(!skipRest) {
+            log.info("Checking for differences between tag terms in names and tags.");
+            List<String> inNamesMissingFromTags = s.compareTagTermsInNamesMissingFromTags(
+                    tagTermsFromNames,
+                    tagTerms
+            );
+            List<String> inTagsMissingFromNames = s.compareTagTermsInTagsMissingFromNames(
+                    tagTerms,
+                    tagTermsFromNames
+            );
+
+            // If there is a name with [TAGY], but no tag "TAGY", then throw warning but add TAGY to the list of tags.
+            // Currently, another validator process in NamesValidator calls Substance.addLocator. This adds the tag.
+            // Actually does not seem to get added by that process ... so including this for now, discuss with Tyler/Larry
+            if (!inNamesMissingFromTags.isEmpty()) {
+                if (addTagWhenInTagsPresentInNames == false) {
+                    log.info("Processing inNamesMissingFromTags.");
+                    GinasProcessingMessage mes = GinasProcessingMessage
+                            .WARNING_MESSAGE("These tag terms found in substance names are not present in substance tags:" + inNamesMissingFromTags.toString() + ".");
+                    callback.addMessage(mes);
+                } else if(addTagWhenInTagsPresentInNames)  {
```

Comment:
No need for else if, just else.

Reply:
But I would reverse the logic here ... do one thing if true, and have else be the false case.